# <center>CS 4661 Homework3_Question2</center>

#### <div align = "right">Hiroko Itagaki    (CIN: 305064830) </div>

#### <div align = "right"> Yosep Kim    (CIN: 220612403) </div>

# Question2:	Debt	Prediction

### a- Read	the	dataset	file	“Credit.csv”	(you	should	download	it	from	CSNS),	and	assign	it	to	a	
	

In [1]:
import pandas as pd
import numpy as np
from sklearn import metrics
from sklearn.cross_validation import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import accuracy_score 

# errror whit using this, used model_selection instead
from sklearn.cross_validation import cross_val_score
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split



a = pd.read_csv('Credit.csv')
        

/home/cowboyuniverse/anaconda3/lib/python3.6/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


### b- Check	out	the	dataset.	The	“Credit”	dataset	includes	“balance”	column	(average	credit	 card	debt	for	a	number	of	individuals)	as	target,	as	well	as	several	features:	age,	cards (number	of	credit	cards),	education	(years	of	education),	income	(in	thousands	of	dollars), limit	(credit	limit),	marital	status,	and	rating	(credit	rating).	

In [2]:
a[::15]

Income  Limit  Rating  Cards  Age  Education  Married  Balance
0     14.891   3606     283      2   34         11        1      333
15    20.089   2525     200      3   57         15        1        0
30    34.142   5666     413      4   47          5        1      863
45    77.380   7569     564      3   50         12        1      997
60    35.510   5198     364      2   35         20        0      631
75    33.017   3180     224      2   28         16        1       29
90    20.191   5767     431      4   42         16        1     1023
105   12.031   3182     259      2   58         18        1      210
120   27.241   1402     128      2   67         15        1        0
135   44.473   3500     257      3   81         16        0        8
150   63.931   5728     435      3   28         14        1      581
165   25.383   4527     367      4   46         11        1      570
180   10.635   3584     294      5   69         16        1      423
195   23.883   5384     398      2   73         16        1      802
210   24.543   3206     243      2   62         12        1       95
225   15.079   5673     411      4   28         15        1     1075
240   29.705   3351     262      5   71         14        1      148
255   58.063   4221     304      3   50          8        0      118
270   15.866   3085     217      1   39         13        0      136
285   18.967   1626     156      2   41         11        1        0
300   21.786   4632     355      1   50         17        1      580
315   14.479   3907     296      3   43         16        1      463
330   20.405   4543     329      2   72         17        0     1054
345   30.012   1511     137      2   33         17        1        0
360   53.566   5891     434      4   82         10        0      712
375   20.974   5673     413      5   44         16        1     1000
390  135.118  10578     747      3   81         15        1     1393

### c- Generate	the	feature	matrix	and	target	vector	(target	is	“balance”	in	this	dataset).	Then, normalize	(scale)	the	features	(note:	don’t	normalize	the	target	vector!).	

In [3]:
feature_cols = ['Income', 'Limit' ,'Rating', 'Cards' ,'Age' ,'Education' ,'Married' ]
for f  in feature_cols:
    max_feature = a[f].max()
    a[f]= a[f]/max_feature
X= a[feature_cols]
X.head()

Income     Limit    Rating     Cards       Age  Education  Married
0  0.079787  0.259182  0.288187  0.222222  0.346939       0.55      1.0
1  0.568090  0.477611  0.491853  0.333333  0.836735       0.75      1.0
2  0.560418  0.508517  0.523422  0.444444  0.724490       0.55      0.0
3  0.797947  0.683102  0.693483  0.333333  0.367347       0.55      0.0
4  0.299420  0.351973  0.363544  0.222222  0.693878       0.80      1.0

In [4]:
y = a['Balance']
# y.head()

### d- Split	 the	 dataset	 into	 testing	 and	 training	 sets	 with	 the	 following	 parameters: test_size=0.2,	random_state=2.	

In [5]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2) 

# print(X_train.shape)
# print(y_train.shape)

# print(X_test.shape)
# print(y_test.shape)

### e- Use	Linear	Regression	to	train	a	linear	model	on	the	training	set.	Check	the	coefficients	of	 the	linear	regression	model.	Which	feature	is	the	most	important?	Which	feature	is	the	 least	important?		

In [6]:
my_linreg = LinearRegression()
my_linreg.fit(X_train, y_train)
# printing Theta0 using attribute "intercept_":
print(my_linreg.intercept_)

-451.390136282


In [7]:
linear_regression = my_linreg.coef_
feature_cols

['Income', 'Limit', 'Rating', 'Cards', 'Age', 'Education', 'Married']

In [8]:
import numpy as np
features = np.array(feature_cols)
lr = np.array(linear_regression)
my_linreg_df = pd.DataFrame({'features':features, 'my_linreg':lr})

my_linreg_df

# my_linreg_df.max()
# my_linreg_df.min()

features    my_linreg
0     Income -1413.472225
1      Limit  1357.754109
2     Rating  2441.300429
3      Cards   103.525487
4        Age  -102.024335
5  Education    40.340443
6    Married   -48.455226

### The most important feature is Rating  2441.3 and Least is Age -1413.47

### f- Predict	“balance”	for	the	users	in	testing	set.	Then,	compare	the	predicted	balance	with the	actual	balance	by	calculating	and	reporting	the	RMSE	(as	we	saw	in	lab	tutorial	4).

In [9]:
y_prediction = my_linreg.predict(X_test)
# print(y_prediction)


In [10]:
from sklearn import metrics
import numpy as np

mse = metrics.mean_squared_error(y_test, y_prediction)
rmse = np.sqrt(mse) 

print( 'Root Mean Square Error is ' + str(rmse))

Root Mean Square Error is 154.410746767


### g- Now,	use	10-fold	Cross-Validation	to	evaluate	the	performance	of	a	linear	regression	in	predicting	the	balance.	Thus,	rather	than	splitting	the	dataset	into	testing	and	training, use	Cross-Validation	to	evaluate	the	regression	performance.	What	is	the	RMSE	when	you	use	cross	validation?	

In [11]:
accuracy_list = cross_val_score(my_linreg, X, y, cv=10, scoring='neg_mean_squared_error')
# print(accuracy_list)from sklearn.cross_validation import cross_val_score
# from sklearn.model_selection import cross_val_score

mse_list_positive = -accuracy_list
rmse_list = np.sqrt(mse_list_positive)
# print(rmse_list)

print(rmse_list.mean())

160.331989107


### When we use cross validation, RMSE is 160.331989107